File: src/commands/ping.rs
use poise::serenity_prelude::CreateEmbed;
use poise::{command, CreateReply};
use std::time::{Duration, Instant};

type Error = Box<dyn std::error::Error + Send + Sync>;
type Context<'a> = poise::Context<'a, crate::Data, crate::Error>;

#[command(
  slash_command,
  description_localized("en-US", "Test the bot's latency")
)]
pub async fn ping(ctx: Context<'_>) -> Result<(), Error> {
  let ping_embed = CreateReply::default().embed(
    CreateEmbed::default()
      .title("🌐 Ping!")
      .description("Pinging..."),
  );
  let start = Instant::now();
  let response = ctx.send(ping_embed).await?;
  let finish = start.elapsed();

  // Wait for 1 second so the user can see the "Pinging..." message
  tokio::time::sleep(Duration::from_secs(1)).await;

  response
    .edit(
      ctx,
      CreateReply::default().embed(
        CreateEmbed::default()
          .title("🏓 Pong!")
          .description(format!("Latency: {} ms", finish.as_millis())),
      ),
    )
    .await?;
  Ok(())
}

File: src/commands/ping.ts
import { SlashCommandBuilder, ChatInputCommandInteraction, EmbedBuilder } from 'discord.js'
import { SlashCommand } from '../models/commands'
import { ColorProvider } from '../utils/color-provider'

export const createPingCommand = (colorProvider: ColorProvider): SlashCommand => ({
  command: new SlashCommandBuilder()
    .setName('ping')
    .setDescription('Test the bot\'s latency'),

  async execute (interaction: ChatInputCommandInteraction) {
    const pingEmbed = new EmbedBuilder()
      .setColor(colorProvider.getWarningColor())
      .setTitle('🌐 Ping!')
      .setDescription('Pinging...')

    const start = Date.now()
    await interaction.reply({ embeds: [pingEmbed] })
    const finish = Date.now() - start

    // Wait for 1 second so the user can see the "Pinging..." message
    await new Promise(resolve => setTimeout(resolve, 1000))

    const pongEmbed = new EmbedBuilder()
      .setColor(colorProvider.getSuccessColor())
      .setTitle('🏓 Pong!')
      .setDescription(`Latency: ${finish} ms`)

    await interaction.editReply({ embeds: [pongEmbed] })
  }
})
